### PR TITLE
Dynamic radius adjustments

### PR DESCRIPTION
**Pull request by KonradStaniec**
Add dynamic radius adjustments based on data deleted from db.

Changes introduced:
- Added config option to enable either static or dynamic radius.
- Added using information from database pruning result to adjust radius.
- Current approach is to set radius to `max(furthestNonDeletedElementDistance, fractionOfdDeletedElems * currentRadius)` this way we avoid situation in which we have super close elements in database and our radius drops significantly making it hard to ingest new data.

open quesiton:
- should be initiate some ping/pong action with our peers to inform them about a change or should we just wait until it propagate naturally (through revalidateLoop). For now second apporach is taken.

### CODE REVIEW

**Review by kdeme — COMMENTED** [review]
I'd think the static radius is most important to fix still (no pruning ever)

> 
> open quesiton:
> 
>     * should be initiate some ping/pong action with our peers to inform them about a change or should we just wait until it propagate naturally (through revalidateLoop). For now second apporach is taken.

This is similar to ENR updates in discv5. We currently do not ping actively nodes there. Some graduations possible there:
- Ping all nodes in routing table
- Ping all nodes that have an active session (no new handshakes)
- Ping set of closest nodes.

File: fluffy/network/wire/portal_protocol.nim [hunk 1]
```diff
@@ -420,18 +421,36 @@ proc processContent(
...
+proc getInitialRadius(rc: RadiusConfig): UInt256 = 
+  case rc.kind
+  of Static:
+    return UInt256.fromLogRadius(rc.logRadius)
+  of Dynamic:
+    # In case of dynamic radius we start from maxium value possible to quicky
+    # gather as much data as possible, and also make sure each data piece in
+    # database is in our range after node restart.
+    # Alternative would be to store node radius in database, and intiizalize it
+    # from database after restart
```
[the remark this answers is not in the document]
Comment (KonradStaniec) [reply fluffy/network/wire/portal_protocol.nim:437]: tbh at this point in time I am not sure which option is better.

Definitely storing radius in db is more precise, but on the other hand it can lead to some tricky questions like `what if node is started with static radius, should we then update radius in db ?`. 

Having it starting always from max value  in dynamic case, make it less precise but as I see it (maybe I am missing something) node will just quickly ingest new data and update its radius accordingly. Also it is just simpler in that regards, that after restart with dynamic setting, we are always sure that all elements in database are within our radius without any hassle.

File: fluffy/network/wire/portal_protocol.nim [hunk 2]
```diff
@@ -1065,6 +1085,61 @@ proc neighborhoodGossip*(
...
+  # all keys in database are always in node range hold.
+  # TODO current silent assumption is that both contentDb and portalProtocol are
+  # using the same xor distance function
+  if p.inRange(key):
+    let res = p.contentDB.put(key, content, p.baseProtocol.localNode.id)
+    case res.kind
+    of ContentStored:
+      discard
+    of DbPruned:
+      if p.radiusConfig.kind == Dynamic:
```
[the remark this answers is not in the document]
Comment (KonradStaniec) [reply fluffy/network/wire/portal_protocol.nim:1135]: I considered it, but my main concern was that it can lead to exhausting node storage memory. Even in our test fleet setup when we have multiple nodes running on single machine, if each one of them will have infinite data storage it can lead to some problems when we throw more data at it. 

I think we need to answer question what is main use case for static radius setting, and as I see it is either testing when we want to have nice mesh of peers which will ingest predictable amounts of data or maybe creating some super nodes in the network which will set it radius to max value and ingest every data it can. It both of those cases constraining storage useage is valuable thing.

Also having static radius set statically always kinda mean that you have node with invalid radius. To make it valid, one would need to known total size of data stored in the network , and make informed decision that he would like to store `1/4` of all data (by setting radius to `1/4` of max) which is xxx bytes. In that case, `totalStorageSize` option would really be unnecessary and pruning could be disabled, but imo it is a bit unrealistic.

so tldr:
- static radius option is kinda special and used in several special contexts (thats why I set default as Dynamic)
- constraining storage size is still useful in those contexts (thats why I left pruning on even with static radius set)
- radius is correlated with how much data node wants to store, but its not perfect as size of total data in network is unknown and it constantly changing, so it is still nice to constrain storage useage at node config level

File: fluffy/network/wire/portal_protocol.nim [hunk 3]
```diff
@@ -1065,6 +1085,61 @@ proc neighborhoodGossip*(
       let req = OfferRequest(dst: node, kind: Direct, contentList: contentList)
       await p.offerQueue.addLast(req)
 
+proc adjustRadius(
```
[the remark this answers is not in the document]
Comment (KonradStaniec) [reply fluffy/network/wire/portal_protocol.nim:1088]: yep, agreed on that.  I am not sure we will need it at this point, but I do not discard this possibility.

File: fluffy/network/wire/portal_protocol.nim (hunk 2 above)
[the remark this answers is not in the document]
Comment (kdeme) [reply fluffy/network/wire/portal_protocol.nim:1135]: Selecting a static radius is exactly that, store everything that falls within that radius. This is required in order to setup a testnet with nodes where we will know how much of the total data they will store.
Doing the same thing with the storage option is impossible currently as data is still being added thus the total data continuously changes (and it will do so forever).

When selecting static radius the node operator takes the responsiblity of keeping track of storage of the system. This is fine (it is the same behavior for a current full node). I do agree that dynamic radius should be the default though.
Pruning with static radius will also give issues on protocol layer. Nodes not having data while others expect it. Nodes will accept offers, only to prune the data immediately after, etc.

Just keep them separate as functionality I'd say:
- Dynamic -> Pruning, adjusting radius
- Static -> No pruning, fixed radius, more for testing / dev setup, but could become eventually a version for full nodes / half full nodes or something similar. I'd be fine with putting this option as `hidden` by the way.
